src/chain.js

```diff
@@ -115,84 +106,65 @@ module.exports = (async () => {
     chainId
   ) {
     let capabilityCallIndex = 0;
-    let chainReport = "";
 
     if (!messages.length) {
-      logger.warn(`${chainId} - Empty Message Chain`);
+      logger.error(`[${chainId}] Empty message chain.`);
       return [];
     }
 
-    do {
-      capabilityCallIndex++;
-      const lastMessage = messages[messages.length - 1];
-      if (!lastMessage) {
-        logger.warn("Last Message is undefined");
-        return messages;
-      }
-      if (!lastMessage.content) {
-        logger.warn("Last Message content is undefined");
-        return messages;
-      }
+    logger.info(
+      `[${chainId}] Processing message chain with ${messages.length} messages.`
+    );
 
-      logger.info(
-        `${chainId} - Capability Call ${capabilityCallIndex} started: ${lastMessage.content.slice(
-          0,
-          2400
-        )}...`
-      );
+    try {
+      for (const message of messages) {
+        capabilityCallIndex++;
+        if (!message || !message.content) {
+          logger.error(
+            `[${chainId}] Message or content is undefined. Aborting.`
+          );
+          return messages;
+        }
+
+        logger.info(
+          `[${chainId}] Capability call ${capabilityCallIndex} started. Last message content: ${message.content}...`
+        );
```

Comment:
@rajivsinclair I don't think this part will work:
- The current code processes the last message (which is the user message), while this code snippet loops from first to last, meaning preambled messages will be processed too.
- This code tries to reassign messages = updatedMessages with the intention of processing the new response from LLM (or the result after running a capability). However, `for (const message of messages)` won’t capture this new response message